gbpservice/neutron/nsf/proxy/proxy.py

```diff
@@ -0,0 +1,223 @@
+        sock = socket.socket()
+        print >>sys.stderr, 'connecting to %s port %s' % self.server
+        sock.settimeout(self.conf.connect_max_wait_timeout)
+        try:
+            sock.connect(self.server)
+        except socket.error, exc:
+            print "Caught exception socket.error : %s" % exc
+            return sock, False
+        return sock, True
+
+
+class Connection(object):
+
+    def __init__(self, conf, socket):
+        self._socket = socket
+        self._idle_wait = conf.idle_min_wait_timeout
+        self._idle_timeout = conf.idle_max_wait_timeout
+        self._idle_count_max = (self._idle_timeout / self._idle_wait)
+        self._idle_count = 0
+
+    def _tick(self):
+        self._idle_count += 1
+
+    def _timedout(self):
+        if self._idle_count > self._idle_count_max:
+            raise ConnectionTimedOut
+
+    def idle(self):
+        self._tick()
+        self._timedout()
+
+    def idle_reset(self):
+        self._idle_count = 0
+
+    def recv(self):
+        self._socket.settimeout(self._idle_wait)
+        try:
+            data = self._socket.recv(16)
+            if data and len(data):
+                self.idle_reset()
+                return data
+            self.idle()
+        except socket.timeout:
+            self.idle()
+        return None
+
+    def send(self, data):
+        self._socket.send(data)
+
+    def close(self):
+        self._socket.close()
+
+    def identify(self):
+        return self._socket.fileno()
+
+
+class ProxyConnection(object):
+
+    def __init__(self, conf, unix_socket, tcp_socket):
+        self._unix_conn = Connection(conf, unix_socket)
+        self._tcp_conn = Connection(conf, tcp_socket)
+
+    def close(self):
+        self._unix_conn.close()
+        self._tcp_conn.close()
+
+    def _proxy(self, rxconn, txconn):
+        data = rxconn.recv()
+        if data:
+            txconn.send(data)
+
+    def run(self):
+        try:
+            self._proxy(self._unix_conn, self._tcp_conn)
+            self._proxy(self._tcp_conn, self._unix_conn)
+            return True
+        except:
+            self._unix_conn.close()
+            self._tcp_conn.close()
+            return False
+
+    def identify(self):
+        return '%d:%d' % (
+            self._unix_conn.identify(),
+            self._tcp_conn.identify())
+
+
+class Worker(object):
+
+    def run(self):
+        while True:
+            try:
+                pc = ConnQ.get()
+                if pc.run():
+                    ConnQ.put(pc)
+            except Empty:
+                pass
+            time.sleep(0)
+
+
+class Proxy(object):
+
+    def __init__(self, conf):
+        self.conf = conf
+        # Be a server and wait for connections from the client
+        self.server = UnixServer(conf, self)
+        self.client = TcpClient(conf, self)
+
+    def start(self):
+        for i in range(self.conf.worker_threads):
+            t = threading.Thread(target=Worker().run)
```

Comment:
We may not need multiple process here - because cpu bound logic is less.. mostly it is I/O so threads should be sufficient to handle multiple connections.